configure.ac

```diff
@@ -3439,84 +3448,109 @@ AC_DEFINE_UNQUOTED(MPIR_Ucount,unsigned $MPI_COUNT,[MPIR_Ucount is an unsigned M
-PAC_SET_MPI_TYPE(2d, MPI_INTEGER1, $pac_cv_f77_sizeof_integer1)
-PAC_SET_MPI_TYPE(2e, MPI_COMPLEX4, $pac_cv_f77_sizeof_real2, 2)
-PAC_SET_MPI_TYPE(2f, MPI_INTEGER2, $pac_cv_f77_sizeof_integer2)
-PAC_SET_MPI_TYPE(30, MPI_INTEGER4, $pac_cv_f77_sizeof_integer4)
-PAC_SET_MPI_TYPE(31, MPI_INTEGER8, $pac_cv_f77_sizeof_integer8)
-PAC_SET_MPI_TYPE(32, MPI_INTEGER16, $pac_cv_f77_sizeof_integer16)
-PAC_SET_MPI_TYPE(33, MPIR_CXX_BOOL, $ac_cv_sizeof_bool)
-PAC_SET_MPI_TYPE(34, MPIR_CXX_FLOAT_COMPLEX, $ac_cv_sizeof_Complex)
-PAC_SET_MPI_TYPE(35, MPIR_CXX_DOUBLE_COMPLEX, $ac_cv_sizeof_DoubleComplex)
-PAC_SET_MPI_TYPE(36, MPIR_CXX_LONG_DOUBLE_COMPLEX, $ac_cv_sizeof_LongDoubleComplex)
-PAC_SET_MPI_TYPE(37, MPI_INT8_T, 1)
-PAC_SET_MPI_TYPE(38, MPI_INT16_T, 2)
-PAC_SET_MPI_TYPE(39, MPI_INT32_T, 4)
-PAC_SET_MPI_TYPE(3a, MPI_INT64_T, 8)
-PAC_SET_MPI_TYPE(3b, MPI_UINT8_T, 1)
-PAC_SET_MPI_TYPE(3c, MPI_UINT16_T, 2)
-PAC_SET_MPI_TYPE(3d, MPI_UINT32_T, 4)
-PAC_SET_MPI_TYPE(3e, MPI_UINT64_T, 8)
-PAC_SET_MPI_TYPE(3f, MPI_C_BOOL, $ac_cv_sizeof__Bool)
-PAC_SET_MPI_TYPE(40, MPI_C_FLOAT_COMPLEX, $ac_cv_sizeof_float__Complex)
-PAC_SET_MPI_TYPE(41, MPI_C_DOUBLE_COMPLEX, $ac_cv_sizeof_double__Complex)
-PAC_SET_MPI_TYPE(42, MPI_C_LONG_DOUBLE_COMPLEX, $ac_cv_sizeof_long_double__Complex)
-PAC_SET_MPI_TYPE(43, MPI_AINT_DATATYPE, $MPI_SIZEOF_AINT)
-PAC_SET_MPI_TYPE(44, MPI_OFFSET_DATATYPE, $MPI_SIZEOF_OFFSET)
-PAC_SET_MPI_TYPE(45, MPI_COUNT_DATATYPE, $MPI_SIZEOF_COUNT)
-PAC_SET_MPI_TYPE(46, MPIX_C_FLOAT16, 2)
+AC_DEFINE_UNQUOTED([MPIR_CHAR_INTERNAL],          [MPIR_INT8], [Internal type for MPI_CHAR])
+AC_DEFINE_UNQUOTED([MPIR_UNSIGNED_CHAR_INTERNAL], [MPIR_UINT8], [Internal type for MPI_UNSIGNED_CHAR])
+len=`expr $ac_cv_sizeof_short \* 8`
+AC_DEFINE_UNQUOTED([MPIR_SHORT_INTERNAL],         [MPIR_INT$len], [Internal type for MPI_SHORT])
+AC_DEFINE_UNQUOTED([MPIR_UNSIGNED_SHORT_INTERNAL],[MPIR_UINT$len], [Internal type for MPI_UNSIGNED_SHORT])
+len=`expr $ac_cv_sizeof_int \* 8`
+AC_DEFINE_UNQUOTED([MPIR_INT_INTERNAL],           [MPIR_INT$len], [Internal type for MPI_INT])
+AC_DEFINE_UNQUOTED([MPIR_UNSIGNED_INTERNAL],      [MPIR_UINT$len], [Internal type for MPI_UNSIGNED])
+len=`expr $ac_cv_sizeof_long \* 8`
+AC_DEFINE_UNQUOTED([MPIR_LONG_INTERNAL],          [MPIR_INT$len], [Internal type for MPI_LONG])
+AC_DEFINE_UNQUOTED([MPIR_UNSIGNED_LONG_INTERNAL], [MPIR_UINT$len], [Internal type for MPI_UNSIGNED_LONG])
+len=`expr $ac_cv_sizeof_long_long \* 8`
+AC_DEFINE_UNQUOTED([MPIR_LONG_LONG_INT_INTERNAL], [MPIR_INT$len], [Internal type for MPI_LONGLONG_INT])
+len=`expr $ac_cv_sizeof_float \* 8`
+AC_DEFINE_UNQUOTED([MPIR_FLOAT_INTERNAL],         [MPIR_FLOAT$len], [Internal type for MPI_FLOAT])
+len=`expr $ac_cv_sizeof_double \* 8`
+AC_DEFINE_UNQUOTED([MPIR_DOUBLE_INTERNAL],        [MPIR_FLOAT$len], [Internal type for MPI_DOUBLE])
+len=`expr $ac_cv_sizeof_long_double \* 8`
+AC_DEFINE_UNQUOTED([MPIR_LONG_DOUBLE_INTERNAL],   [MPIR_ALT_FLOAT$len], [Internal type for MPI_LONG_DOUBLE])
```

Comment:
Hard code `long double` to `MPIR_ALT_FLOAT{12,16}` here, assuming `long double` is not `__float128`. 

FIXME: the assumption may not be true for some compilers. We need add configure test for the actual floating format.  I am kicking the can down the road until some one report the issue.

Reply:
Indeed, the GCC manual page is laughing in our faces :rofl: :rofl: :rofl: with option like the following:
```
-m128bit-long-double
-m96bit-long-double
-mlong-double-64
-mlong-double-80  
-mlong-double-128
```

Reply:
We also hard-define `MPIR_ALT_FLOAT{12,16}` to `long double` (see comment below), so we are good either way. We can keep adding internal `_ALT_` types to deal with any odd types, but hopefully, there won't me many. :)